src/sentry/api/bases/external_actor.py

```diff
@@ -0,0 +1,120 @@
+from typing import Any, MutableMapping
+
+from django.db import IntegrityError
+from django.http import Http404
+from rest_framework import serializers
+from rest_framework.exceptions import PermissionDenied
+
+from sentry import features
+from sentry.api.serializers.rest_framework.base import CamelSnakeModelSerializer
+from sentry.api.validators.integrations import validate_provider
+from sentry.models import ExternalActor, Organization, OrganizationMember, Team, User
+from sentry.types.integrations import EXTERNAL_PROVIDERS
+
+
+class ExternalActorSerializerBase(CamelSnakeModelSerializer):  # type: ignore
+    external_id = serializers.CharField()
+    external_name = serializers.CharField(required=True)
+    provider = serializers.ChoiceField(choices=list(EXTERNAL_PROVIDERS.values()))
+
+    @property
+    def organization(self):
+        return self.context["organization"]
+
+    def get_actor_id(self, validated_data: MutableMapping[str, Any]) -> int:
+        target_option = validated_data.pop(self._actor_key, None)
+
+        if not target_option:
+            raise Exception("Invalid actor")
+
+        return target_option.actor_id
+
+    def get_provider_id(self, validated_data: MutableMapping[str, Any]) -> int:
+        provider_name_option = validated_data.pop("provider", None)
+        return validate_provider(provider_name_option).value
+
+    def create(self, validated_data: MutableMapping[str, Any]) -> ExternalActor:
+        actor_id = self.get_actor_id(validated_data)
+        provider = self.get_provider_id(validated_data)
+
+        return ExternalActor.objects.get_or_create(
+            **validated_data,
+            actor_id=actor_id,
+            provider=provider,
+            organization=self.organization,
+        )
+
+    def update(
+        self, instance: ExternalActor, validated_data: MutableMapping[str, Any]
+    ) -> ExternalActor:
+        # Discard the object ID passed by the API.
+        if "id" in validated_data:
+            validated_data.pop("id")
+
+        for key, value in validated_data.items():
+            setattr(self.instance, key, value)
+        try:
+            self.instance.save()
+            return self.instance
+        except IntegrityError:
+            raise serializers.ValidationError(
+                "There already exists an external user association with this external_name and provider."
```

Comment:
```suggestion
"There already exists an external actor association with this external_name and provider."
```

Reply:
I think we should keep actors out of user-facing language. Can i say "user or team"?

Reply:
I agree since it's an internal object. Maybe we can use the `_actor_key` to be more specific? Yeah otherwise I'd say something generic like ```"There already exists an external association with this external_name and provider."```